### ISSUE TITLE
v0.9.0 preliminary testing and feedback

### ISSUE THREAD

**Issue by niklasramo** [haltu/muuri#373]
The v0.9.0 release is (hopefully 🤞) just around the corner, but before the release I'd like to ask for anyone interested to take the [`dev`](https://github.com/haltu/muuri/tree/dev) branch for a test drive and just explore it, and try to break it 🔨. It's currently in pretty good shape... I'd even use the word stable 🙂 

For reference, here's the big picture of what has changed since v0.8.0: https://github.com/haltu/muuri/milestone/7, but there are _a lot_ of minor changes here and there.

API has also changed a bit, but I've tried my best to keep the README updated with the changes.

To get you started, here's a codesandbox with the latest `dev` in use: https://codesandbox.io/s/muuri-autoscroll-dev-17lib. You can use this link https://cdn.jsdelivr.net/gh/haltu/muuri@dev/dist/muuri.min.js to get the latest version of dev branch at any time.

All feedback is welcome 🙂

**Comment by Paol-imi** [comment]
### Features that could be changed

Some operations that Muuri has when an item is sent to another grid:

- During the drag in which the item changes the grid, the settings relating to the drag remain relative to the old grid (e.g. dragSortHeuristics). Scroll targets are also not updated.
- Drag events (e.g. dragEnd) are triggered in the old grid.

### Question

Have you changed anything in the way Muuri handles new inactive items added? 

Now when I make the items visible, the positioning animations start from the top left, while before they were already positioned. Maybe some settings have changed that I haven't updated.

**Comment by niklasramo** [comment]
> During the drag in which the item changes the grid, the settings relating to the drag remain relative to the old grid (e.g. dragSortHeuristics). Scroll targets are also not updated.
Drag events (e.g. dragEnd) are triggered in the old grid.

Yep, this is by design. Not saying it's the best design, but to me it made sense that the drag procedure remains always in the scope of the originating grid until to the point you release the item.

> Have you changed anything in the way Muuri handles new inactive items added?

Yes. The async layout feature broke the old behavior so now we can't trust the layout to happen immediately after the items are shown anymore.

Previously when hidden item was shown we made the item visible, refreshed its dimensions and then called the layout. All of this happened synchronously so there was no issues with the hidden item being in incorrect position when we show it, because the user would never see that since the next paint operation happened only after the whole layout procedure was finished, which sets the item's translate values in the end. 

Now, when an item is shown we need to first make it temporarily visible to retrieve its dimensions from the DOM and then hide it again until the next layout is calculated, and then show it along with the next layout process.  This is process is baked into grid.show/hide/filter() methods, which now have a new option `syncWithLayout` (enabled by default) that automatically waits for the next layout to be ready before showing/hiding the items.

The bottom line is, that you have to now be aware all the time that the layout process is _not_ guaranteed to be synchronous. It _might_ be, but it might not, depends on the layout algorithm.

That being said, you _can_ force the layout process to be synchronous by providing your own layout algorithm. You could even replace the default layout algorithm with a new one, which does not use web workers and runs synchronously. It's this simple: `Muuri.defaultPacker = new Muuri.Packer(0);`. The number argument that the `Packer` accepts tells how many web workers should be created for the layout calculations, if it's `0` then everything will run in main thread synchronously.

**Comment by Paol-imi** [comment]
All very clear, amazing work

**Comment by niklasramo** [comment]
One thing I'd like people to test is the TypeScript support, are the types working out alright?

**Comment by Paol-imi** [comment]
I am trying to extend (for now successfully) some definitions of muuri into muuri-react.

For now I have not encountered any problems except [this](https://github.com/haltu/muuri/blob/dev/src/index.d.ts#L157) which I think should return  `boolean | undefined`

**Comment by niklasramo** [comment]
Nice catches! You are very correct about both of those issues, I'll fix them :+1:

**Comment by niklasramo** [comment]
I've been testing the dev version by trying to build a bit more advanced version of the kanban board demo in the website. Found already a couple of hiccups.

1. There's a bit of a problem in all the cases where a classname is removed simultaneously with the element being reparented. CSS transitions do not behave as expected in this scenario and e.g. :hover related transitions are replayed from start after reparenting if the item is being hovered during the reparenting. A pretty simple fix is just letting the classnames linger one tick's duration longer on the element after the classname is removed. This way you can prevent the transitions replaying on during reparenting scenarios.

2. I ended up in a scenario where I wanted to set `display:flex` to an item element, and found really no justification for the `display:block` enforcement. This one I already changed in the dev branch.

3. I found myself in a scenario where I really needed to autoscroll the grid container without using a fixed drag container, and that scenario does not work so well. So I'll need to hammer my head with that one still a bit.
